pulp_rpm/app/advisory.py

```diff
@@ -0,0 +1,206 @@
+from gettext import gettext as _
+from itertools import chain
+
+import hashlib
+
+import createrepo_c as cr
+
+from django.db import (
+    IntegrityError,
+    transaction,
+)
+from django.utils.dateparse import parse_datetime
+
+from pulpcore.plugin.models import (
+    Content,
+    RepositoryContent,
+)
+
+from pulp_rpm.app.exceptions import AdvisoryConflict
+
+
+def resolve_advisories(version, previous_version):
+    """
+    Decide which advisories to add to a repo version and which to remove, and adjust a repo version.
+
+    Args:
+        version (pulpcore.app.models.RepositoryVersion): current incomplete repository version
+        previous_version (pulpcore.app.models.RepositoryVersion):  a version preceding
+                                                                   the current incomplete one
+
+    """
+    content_pks_to_add = set()
+    content_pks_to_remove = set()
+    content_pks_to_exclude = set()  # exclude from the set of content which is being added
+
+    # identify conflicting advisories
+    current_advisories = version.content.filter(pulp_type='rpm.advisory')
+    added_advisories = current_advisories
+    advisory_conflicts = []
+    if previous_version:
+        previous_advisories = previous_version.content.filter(pulp_type='rpm.advisory')
+        previous_advisory_ids = set(previous_advisories.values_list('rpm_updaterecord__id',
+                                                                    flat=True))
+        # diff for querysets works fine but the result is not fully functional queryset,
+        # e.g. filtering doesn't work
+        added_advisories = current_advisories.difference(previous_advisories)
+        if len(list(added_advisories)) != len(set(added_advisories)):
+            raise AdvisoryConflict(_('It is not possible to add two advisories of the same id to '
+                                     'a repository version.'))
+        added_advisory_ids = set(adv.rpm_updaterecord.id for adv in added_advisories)
+        advisory_conflicts = added_advisory_ids.intersection(previous_advisory_ids)
+
+    added_advisory_pks = [adv.rpm_updaterecord.pk for adv in added_advisories]
+    for advisory_id in advisory_conflicts:
+        previous_advisory = previous_advisories.get(
+            rpm_updaterecord__id=advisory_id).rpm_updaterecord
+        added_advisory = version.content.get(
+            rpm_updaterecord__id=advisory_id, pk__in=added_advisory_pks).rpm_updaterecord
```

Comment:
If you convert `previous_advisories` to `UpdateRecords`:

```python
previous_advisories = UpdateRecord.filter(pk__in=previous_version.content.filter(pulp_type='rpm.advisory'))
```

You could simplify these lines:

```python
previous_advisory = previous_advisories.get(id=advisory_id)
added_advisory = version.content.get(id=advisory_id, pk__in=added_advisory_pks)
```

Reply:
Also, does this not work?

```python
added_advisory = version.content.get(id=advisory_id, pk__in=added_advisories)
```

Reply:
@daviddavis, I'll try.
Well, maybe it's the right thing to do everything  with the same approach, so I'll probably convert to use UpdateRecord, similar to how you use Modulemd. I just want travis to pass, and then I cam refactor. Testing of correct merging takes some time as well.

Reply:
> Also, does this not work?
> 
> ```python
> added_advisory = version.content.get(id=advisory_id, pk__in=added_advisories)
> ```

This doesn't work because content is the master model, so it will be `rpm_updaterecord__id=advisory_id` but the main problem is in `added_advisories`.
I get `django.db.utils.ProgrammingError: subquery has too many columns` (either by using `version.content` or `UpdateRecord`). It's all because `added_advisories` is a result of `qs.difference`. So I'm back to collecting `pk`s.